### ISSUE TITLE
Many backwards compatibility tests are broken in 2.x

### ISSUE THREAD

**Issue by nik9000** [elastic/elasticsearch#13522]
I'm working on getting the backwards compatibility tests in 2.x but `BasicBackwardsCompatibilityIT` is proving to be very difficult so I'm pulling it out into a separate issue. This issue.

* [x] SnapshotBackwardsCompatibilityIT (#14845)
* [x] SignificantTermsBackwardCompatibilityIT (#14948)
* [x] FunctionScoreBackwardCompatibilityIT (#15332)
* [x] RecoveryBackwardsCompatibilityIT (#14817)

**Comment by nik9000** [comment]
I'm finding more and more and more backwards compatibility tests failing. Each one is a rabbit hole I can lose several hours to. I'm going to spend no more than an hour on each and if I can't fix it then I'll just mark it AwaitsFix with this bug.

**Comment by nik9000** [comment]
There are a few things `@AwaitsFix`ed on this issue. Also we suppress any `index.data_path` set on the indexes during the BWC tests and we _probably_ shouldn't have to do that. We should figure that out as part of this issue as well.

**Comment by nik9000** [comment]
Note for those following along at home - I don't believe this should block 2.1.0. Its fine to kick this to 2.1.1, 2.1.2, etc. The *really* important BWC tests are passing.

**Comment by nik9000** [comment]
I'm working on SnapshotBackwardsCompatibilityIT on and off now.

**Comment by nik9000** [comment]
L

**Comment by nik9000** [comment]
misclick

**Comment by nik9000** [comment]
> FunctionScoreBackwardCompatibilityIT

I got a start on this but it'll be a huge pain to do until we bundle the tar plugin in the distribution. The trouble is the security policy. Right now the bwc tests run with our default security policy which means they can't run groovy code and neither can the embedded nodes. If we bundled groovy then it'd be _fairly_ simple to have the test run only external nodes. In fact the bwc tests should always be running only external nodes and this is a good excuse to do that across the board.

**Comment by nik9000** [comment]
As of now all of the scary backwards compatibility tests have been fixed for 2.2. The only remaining one is that `FunctionScoreBackwardCompatibilityIT` which so far as I know really wants to make sure groovy works across a mixed version cluster. I don't think this is a huge, huge issue, but it needs fixing eventually.

**Comment by nik9000** [comment]
This is probably the biggest remaining BWC task:
https://github.com/elastic/elasticsearch/issues/14406

**Comment by brwe** [comment]
> The only remaining one is that FunctionScoreBackwardCompatibilityIT which so far as I know really wants to make sure groovy works across a mixed version cluster.

It was supposed to test that old nodes can parse whatever new nodes send. I added the script then because I thought we might want to check that as well but for now we can simply remove that. I made a pr here: #15332 

>  I don't think this is a huge, huge issue, but it needs fixing eventually.

I too think the problem with testing scripting in mixed clusters remains. We currently cannot test that old nodes can parse requests related to scripting that were sent by new nodes. I do wonder though if there is a more clever way to verify that parsing on old nodes still works without having to spin up a cluster and send requests...

**Comment by brwe** [comment]
Maybe we could make all nodes (old and new) start the same way with the external node service instead of using node builder etc. This would have the advantage that the cluster would be an actual cluster. The test would only start node clients or transport clients that then connect to the cluster. Then we could use whatever modules are available per default and also install whatever plugins we want before we start the test. Would be a larger change though.

**Comment by nik9000** [comment]
> Maybe we could make all nodes (old and new) start the same way with the external node service instead of using node builder etc.

Yeah! I think it is a good idea too. Maybe another meta issue for improvements in BWC tests?

**Comment by nik9000** [comment]
Britta's fixed the last test! Hurray! Closing this tracking issue.
